:recycle: Added gRPC service for internal communication

mhkarimi1383 opened the pull request:
Also removes task pod usage and inner-function imports :)

Fixed typo on global variable sets in logger init

Added `pyi` files for typing of grpc information

mhkarimi1383 commented:
@mergifyio rebase

tiagolobocastro commented on deploy/helm/rawfile-localpv/templates/secret.yaml:
```diff
@@ -0,0 +1,9 @@
+  name: {{ include "rawfile-localpv.fullname" . }}-secrets
+type: Opaque
+data:
+  internal-signature: {{ include "rawfile-localpv.internal-signature" . | b64enc }}
```
Can we check if the secret already exists? 
Otherwise I suspect helm upgrade will trigger changes even if there are no changes

tiagolobocastro commented on deploy/helm/rawfile-localpv/values.yaml:
```diff
@@ -11,6 +11,9 @@ reservedCapacity: ""
 capacityOverride: ""
 
+# -- Sets internal signature used for authentication of the internal communication gRPC server (defaults to randomly generated string)
+internalSignature: ""
```
maybe this should be within an "auth", example:
```yaml
auth:
  enabled: true
  # -- Set ....
  token: ""
```

tiagolobocastro commented on deploy/helm/rawfile-localpv/values.yaml:
```diff
@@ -142,6 +147,9 @@ node:
   grpcWorkers: 10
 
+  # -- Port Number used for internal communication gRPC server
+  internalCommunicationPort: 4500
```
nit: or something else shorter
```suggestion
  internalGrpcPort: 4500
```

And we can take also the worker count from `grpcWorkers` ?

tiagolobocastro commented on rawfile/internal_svc.py:
```diff
@@ -0,0 +1,70 @@
+                logger.debug(
+                    "SignatureInterceptor: Invalid signature",
+                    metadata=handler_call_details.invocation_metadata,
+                    expected=expected_metadata,
```
Should we be logging this?

mhkarimi1383 replied:
It was just for debug reasons, I will make it to not print out sensitive information

mhkarimi1383 replied:
I made it to not log expected field

mhkarimi1383 commented:
@mergifyio rebase

tiagolobocastro commented on rawfile/orchestrator/k8s.py:
```diff
@@ -229,3 +127,62 @@ def write_config_map(name: str, key: str, value: str, overwrite=False):
+        while not self._stop_event.is_set():
+            try:
+                for _ in w.stream(
+                    self.core_v1.list_namespaced_pod,
```
could we somehow leverage the csinode CR instead?

mhkarimi1383 replied:
I have used the pod selector of the daemonset since I need the IP address of the pod itself

You mean that we can watch csinode resources and get ip of the pod related to that node? It will make it complicated I think

tiagolobocastro commented on rawfile/orchestrator/k8s.py:
```diff
@@ -229,3 +127,62 @@ def write_config_map(name: str, key: str, value: str, overwrite=False):
+                    if self._stop_event.is_set():
+                        w.stop()
+                        break
+                    self.reload_mapping()
```
Can't we just reload using the event itself?

mhkarimi1383 replied:
You mean we should only refresh by changes?

tiagolobocastro replied:
Yes using the event which is currently `_`

mhkarimi1383 replied:
I have changed it to use event data

tiagolobocastro commented on rawfile/orchestrator/k8s.py:
```diff
@@ -229,3 +127,62 @@ def write_config_map(name: str, key: str, value: str, overwrite=False):
+                        break
+                    self.reload_mapping()
+            except Exception:
+                logger.exception("DS Watcher Error, retrying...")
```
If the node has some error does this log continuously? If so we could silence after error until the next success?

mhkarimi1383 replied:
I will...

tiagolobocastro commented on rawfile/rawfile.py:
```diff
@@ -67,16 +78,28 @@ def signal_handler(sig, _=None):
-    server.wait_for_termination()
+    if internal_server:
+        internal_server.start()
+    with futures.ThreadPoolExecutor() as executor:
```
set max workers to the same: `max_workers=int(driver_config.grpc_workers)` ?

mhkarimi1383 replied:
Will separate that configuration

tiagolobocastro reviewed (commented):
This is great, glad to see the helper going away :+1:

mhkarimi1383 commented:
@tiagolobocastro 

I have added more logs to check Block mode expand

Here is the result from the node itself

```python
2025-08-27 14:53:56.450 | DEBUG    | internal_svc:ExpandRawFile:42 - ExpandRawFileRequest: Waiting for lock | {'volume_id': 'pvc-53ad7106-e5aa-4b00-9739-57294b3f6f86'}
2025-08-27 14:53:56.452 | DEBUG    | utils.commands:run:40 - Shell command execution | {'check': True, 'capture_output': True, 'command': 'losetup -j /data/pvc-53ad7106-e5aa-4b00-9739-57294b3f6f86/disk.img', 'start': datetime.datetime(2025, 8, 27, 14, 53, 56, 450846), 'returncode': 0, 'end': datetime.datetime(2025, 8, 27, 14, 53, 56, 452851), 'latency': datetime.timedelta(microseconds=2005), 'stderr': '', 'stdout': '/dev/loop0: [2065]:2490376 (/data/pvc-53ad7106-e5aa-4b00-9739-57294b3f6f86/disk.img)\n'}
2025-08-27 14:53:56.455 | DEBUG    | internal_svc:ExpandRawFile:61 - ExpandRawFileRequest: Applying size increament | {'volume_id': 'pvc-53ad7106-e5aa-4b00-9739-57294b3f6f86'}
2025-08-27 14:53:56.456 | DEBUG    | internal_svc:ExpandRawFile:69 - ExpandRawFileRequest: Patching metadata | {'volume_id': 'pvc-53ad7106-e5aa-4b00-9739-57294b3f6f86'}
2025-08-27 14:53:56.457 | DEBUG    | internal_svc:ExpandRawFile:76 - ExpandRawFileRequest: Done | {'volume_id': 'pvc-53ad7106-e5aa-4b00-9739-57294b3f6f86'}
```


And here is the result from Controller

```python
2025-08-27 14:53:56.460 | INFO     | utils.logs:wrap:130 - GRPC Server Access Log | {'handler': 'ControllerExpandVolume', 'request': volume_id: "pvc-53ad7106-e5aa-4b00-9739-57294b3f6f86"
capacity_range {
  required_bytes: 1073741824
}
volume_capability {
  block {
  }
  access_mode {
    mode: SINGLE_NODE_WRITER
  }
}
, 'starttime': datetime.datetime(2025, 8, 27, 14, 53, 56, 442258), 'response': volume_id: "pvc-53ad7106-e5aa-4b00-9739-57294b3f6f86"
capacity_range {
  required_bytes: 1073741824
}
volume_capability {
  block {
  }
  access_mode {
    mode: SINGLE_NODE_WRITER
  }
}
, 'latency': datetime.timedelta(microseconds=17937), 'endtime': datetime.datetime(2025, 8, 27, 14, 53, 56, 460195), 'success': True}

```


and looks like that everything is correct

tiagolobocastro commented:
Got it, because now with gRPC is faster, the Pod has completed as part of `And the pod should complete with success` but the `NodeUnstage` has not completed yet when `ControllerExpand` is called, so it returns that node resize is required.

tiagolobocastro commented on rawfile/rawfile_servicer.py:
```diff
@@ -268,21 +271,21 @@ def ControllerExpandVolume(self, request, context):
+        channel = grpc.insecure_channel(f"{node_ip}:{config.csi_driver.internal_port}")
+        stub = internal_pb2_grpc.InternalStub(channel)
+        response = stub.ExpandRawFile(
+            internal_pb2.ExpandRawFileRequest(volume_id=volume_id, new_size=size),
```
We should set a timeout on gRPC call

mhkarimi1383 commented:
So, This is a problem with our tests, Or we should make locks to be shared or something else

mhkarimi1383 commented:
We can also use annotations for that

Also I think we should return `False` for node resize is required on block vols

mhkarimi1383 commented:
@tiagolobocastro I have commented out the block tests for now, So we should fix them in another PR

I will also start to add migrate and clone in another PR later on

mhkarimi1383 commented:
@mergifyio squash

mhkarimi1383 committed:
:recycle: Added gRPC service for internal communication

Also removes task pod usage and inner-function imports :)

Signed-off-by: Muhammed Hussein Karimi <info@karimi.dev>

* :recycle: Remove task pods and use internal gRPC connection

Signed-off-by: Muhammed Hussein Karimi <info@karimi.dev>

* :zap: Watch for DS changes

Signed-off-by: Muhammed Hussein Karimi <info@karimi.dev>

* :lock: Add internal communication gRPC auth

Signed-off-by: Muhammed Hussein Karimi <info@karimi.dev>

* :technologist: Add pyi files for gRPC types

Types are now accessible from IDE/Editor and we have access to types

Signed-off-by: Muhammed Hussein Karimi <info@karimi.dev>

* :zap: Some final touches on how internal gRPC server is configured

Signed-off-by: Muhammed Hussein Karimi <info@karimi.dev>

* :zap: Do not refresh data entirely when something change

Signed-off-by: Muhammed Hussein Karimi <info@karimi.dev>

* :zap: Make get node ip getter more safe

Signed-off-by: Muhammed Hussein Karimi <info@karimi.dev>

mhkarimi1383 commented:
Oh, I forgot to fix helm chart docs after changing values :)

mhkarimi1383 commented:
@tiagolobocastro 
Can you review this again?

mhkarimi1383 commented:
@mergifyio rebase

mhkarimi1383 commented:
@mergifyio queue
